### DISCUSSION TITLE
Inefficient Query being timed out

### DISCUSSION THREAD

**Discussion by perroudsky** [prisma/prisma#8401] [Q&A]
Hi everyone,

I've built some sort of swipe where the user can swipe through a randomised order of product (like tinder).

The query worked just fine until the amount of already swiped products went up (~60 K). Now the query takes longer than 30 seconds and so I'm getting timeout errors on Heroku.

The way i've made it is like that : To not show an already swiped product, each time the user swipe, it is saved as a SwipeNope or SwipeLike in db.

I then query a batch of products (for exemple 20) and make sure in the query that it does not exist as a swipeLike or swipeNope.
The cursor based approach does not work since I need the product order to be randomized and new products are added daily.

Here is how the query look like:

```js
const opArgs = {
    take: args.take,
    orderBy: { randomizer: 'asc' },
    where: {
      AND: [
        { type: 'PRODUCT' },
        { isSearchable: true },
        {
          swipeLikes: {
            every: {
              author: {
                id: {
                  not: args.userId,
                },
              },
            },
          },
        },
        {
          swipeNopes: {
            every: {
              author: {
                id: {
                  not: args.userId,
                },
              },
            },
          },
        },
        { gender: args.swipeFilters.gender },
        { ...rangeArgs },
        { ...targetArgs },
      ],
    },
  }
  const posts = await prisma.post.findMany(opArgs)
  return posts
```

if you comment out the "swipeNopes" object it worked fine so I guess this part is the inefficient one.

Here is a simplified version of how prisma translate it to sql :

```sql
SELECT
	"public"."Post"."id",
FROM
	"public"."Post"
WHERE (("public"."Post"."type" = 'PRODUCT'
		AND "public"."Post"."isSearchable" = TRUE
		AND "public"."Post"."gender" = 'MEN'
		AND("public"."Post"."id")
		NOT IN(
			SELECT
				"t0"."id" FROM "public"."Post" AS "t0"
				INNER JOIN "public"."SwipeLike" AS "j0" ON ("j0"."postId") = ("t0"."id")
			WHERE ((NOT("j0"."id")
				IN(
					SELECT
						"t1"."id" FROM "public"."SwipeLike" AS "t1"
						INNER JOIN "public"."User" AS "j1" ON ("j1"."id") = ("t1"."authorId")
					WHERE ("j1"."id" <> 'ckovo8n76228881mw1vzx1uala'
						AND "t1"."id" IS NOT NULL)))
			AND "t0"."id" IS NOT NULL))
		AND("public"."Post"."id")
		NOT IN(
			SELECT
				"t0"."id" FROM "public"."Post" AS "t0"
				INNER JOIN "public"."SwipeNope" AS "j0" ON ("j0"."postId") = ("t0"."id")
			WHERE ((NOT("j0"."id")
				IN(
					SELECT
						"t1"."id" FROM "public"."SwipeNope" AS "t1"
						INNER JOIN "public"."User" AS "j1" ON ("j1"."id") = ("t1"."authorId")
					WHERE ("j1"."id" <> 'ckovo8n76228881mw1vzx1uala'
						AND "t1"."id" IS NOT NULL)))
			AND "t0"."id" IS NOT NULL))
		

	))
ORDER BY
	"public"."Post"."randomizer" ASC
LIMIT 20
```

So here's my question, any idea on making it more efficient, or maybe any idea of implementation changes ?
thanks!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
